apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs

```diff
@@ -215,5 +217,129 @@ string ExtractPath(DocumentReference documentReference)
         internal IAsyncEnumerable<CollectionReference> ListCollectionsAsync(DocumentReference parent) =>
             Client.ListCollectionIdsAsync(parent?.Path ?? RootPath)
                 .Select(id => new CollectionReference(this, parent, id));
+
+        // TODO: Is it appropriate to use ConfigureAwait(false) here? Should the caller be able to specify the context for executing their callback?
+        // TODO: Collect up all the exceptions we've retried?
+        // TODO: Is it nice to have all these overloads?
+
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        /// <summary>
+        /// Runs a transaction asynchronously, with a synchronous callback that returns a value.
+        /// The specified callback is executed for a newly-created transaction. If committing the transaction
+        /// fails, the whole operation is retried based on <see cref="TransactionOptions.MaxAttempts"/>.
+        /// </summary>
+        /// <typeparam name="T">The result type of the callback.</typeparam>
+        /// <param name="callback">The callback to execute. Must not be null.</param>
+        /// <param name="options">The options for the transaction. May be null, in which case default options will be used.</param>
+        /// <param name="cancellationToken">A cancellation token for the operation. This is exposed to the callback through <see cref="Transaction.CancellationToken"/>
+        /// and applied to all RPCs to begin, rollback or commit the transaction.</param>
+        /// <returns>A task which completes when the transaction has committed. The result of the task then contains the result of the callback.</returns>
+        public Task<T> RunTransactionAsync<T>(Func<Transaction, T> callback, TransactionOptions options = null, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(callback, nameof(callback));
+            return RunTransactionAsync(async transaction => callback(transaction), options, cancellationToken);
```

Comment:
Is there any advantage of `async transaction => ...` to ` transaction => Task.FromResult(...)` ?
The later seems to me to be easier to understand! And won't give CS1998 warning.

Reply:
It was for the sake of propagating exceptions in the same way - but as we're awaiting the result immediately, it should be fine actually. Will change it.